### ISSUE TITLE
Aspect ratio when using 2 displays

### ISSUE THREAD

**Issue by cyrilst** [popcornmix/omxplayer#688]
Hello, I think there is an aspect ratio problem when using 2 displays with 2 different ratios. First let me explain the 2 use cases (in each case, there is no graphical interface, omxplayer is just launched via ssh or cron):

- the first use case works perfectly well: I plug a Raspberry Pi 3B+ onto an HDMI screen (any resolution / aspect ratio up to Full HD). I then launch omxplayer with any kind of media aspect ratio, and the option --aspect-mode values (letterbox, fill, stretch) work as intended.

- the second use cases is a bit more complex: I plug a Raspberry Pi 3B+ onto the same HDMI screen, but I also have the official 7" Raspberry Pi touch screen plugged. I use the option --display 5 in order to display the video on the HDMI screen (I just display a picture with the frame buffer interface on the touchscreen). The touchscreen is 800x480, and has an aspect ratio between 16:9 and 16:10. When I launch omxplayer, the aspect ratio is never good:

  - when the screen is 1920x1080 and the video is also 1920x1080, with the option --aspect-mode letterbox, the video is stetched and there are black bars over and under the picture.

  - this problem persists even if I use the option --win 0,0,1919,1079

  - the only way to have a non stretched video is by explicitly giving a size with the --win option and setting the option --aspect-mode stretch.

In conclusion, it looks like the option --display 5 is almost ok, but is a bit buggy if 2 screens are used and not in the same aspect ratio. Is there a way to get around this problem ? The --win + stretch mode is not really an option, as screens and medias ratios can change regularly…

Thank you.

**Comment by cyrilst** [comment]
One more info: if I set display_default_lcd=0 in /boot/config.txt, the HDMI becomes the main screen, and everything works fine. The aspect ratio is always good on the HDMI screen, and also on the touch screen (using --display 4).

But if I do this, the frame buffer interface is on the HDMI screen… I need it on the touch screen. That's why I need to set display_default_lcd=1 and use --display 5 in omxplayer.

**Comment by popcornmix** [comment]
I suspect the problem is we call:
``vc_tv_get_display_state(&current_tv_state)``
and then we use the ``aspect_ratio`` from ``current_tv_state``.
But that is the default display rather than the display being driven by omxplayer from ``--display``.

I'm not sure if an api is available to query the aspect ratio of the secondary display - I'll need to look into this.

**Comment by cyrilst** [comment]
Thanks for the reply. Indeed it looks like it's the cause. In cli, `tvservice -s` returns only the default display size.

Funny enough, `tvservice --dumpedid` always gives you info about the HDMI screen, whether it's the default display or not. This is logical, but maybe is it worth looking in this direction ?

**Comment by KenT2** [comment]
For me the problem is associated with having a Pi4 with official touchscreen and one HDMI monitor

The HDMI monitor is 1920*1080 16:9

I am playing suits-short.mkv, the video that was used to develop subtitles for omxplayer
It is 1276*720 = 15.95:9

With the touchscreen disabled (ignore_lcd= 1 in config.txt) all is OK
omxplayer --display 2 suits-short.mkv fills the screen
omxplayer -- display 2 --aspect-mode letterbox suits-short.mkv fills the screen as it should do since aspect ratios are the same for video and display (well nearly!)

With the touchscreen enabled there are big black bars at top and bottom of screen
omxplayer --display 2 suits-short.mkv - black bars
omxplayer -- display 2 --aspect-mode letterbox suits-short.mkv - black bars, should definitely fill the screen in letterbox
omxplayer -- display 2 --aspect-mode fill suits-short.mkv fills the screen as expected

Pi3 with a single HDMI display has always been full screen for
omxplayer --display 2 suits-short.mkv

**Comment by dbussert** [comment]
same problem as @KenT2 , rpi4 with official touchscreen and also hdmi out on display 2. A 1920x1080 file playback with omxplayer on a 1920x1080 hdmi display has black bars. The work-around is to use `--apsect-mode stretch` which will fill all of the space. `fill` causes the left and right sides to go off the screen as it fills the vertical space.
